### PR TITLE
Add a ModelingToolkit implementation 

### PR DESCRIPTION
**Pull request by Vaibhavdixit02**
As discussed in https://github.com/SciML/Optimization.jl/issues/537 there were updates to the structural simplification in MTK. The functional interface with Optimization can be switched to using another AD backend (I'll make another PR for that) and this can be used as the MTK version since it is written in the symbolic interface.

### CODE REVIEW

**Comment by odow** [comment]
How does this do on say, the 118 case here: https://github.com/lanl-ansi/rosetta-opf/blob/main/data/pglib_opf_case118_ieee.m?

**Comment by odow** [comment]
On the 118 bus case I get

```Julia
(pm) pkg> st
      Status `/private/tmp/pm/Project.toml`
  [b6b21f68] Ipopt v1.3.0
  [961ee093] ModelingToolkit v8.58.0
  [7f7a1694] Optimization v3.14.2
  [fd9f6733] OptimizationMOI v0.1.13
  [c36e90e8] PowerModels v0.19.9
  [37e2e46d] LinearAlgebra

julia> solve_opf("case118.m")
[info | PowerModels]: removing 3 cost terms from generator 32: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 29: [3266.8781000000004, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 1: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 54: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 2: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 41: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 51: [3504.3401000000003, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 53: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 27: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 42: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 33: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 28: [3478.1778000000004, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 50: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 52: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 26: [1605.6042000000002, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 10: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 24: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 25: [2486.1868, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 23: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 49: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 5: [2498.342, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 31: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 43: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 39: [3407.2633000000005, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 34: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 44: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 17: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 37: [2460.0771999999997, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 47: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 9: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 12: [2222.098, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 20: [2420.2306, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 6: [12458.1564, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 14: [2599.3982, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 7: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 8: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 19: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 4: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 22: [2727.7342999999996, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 11: [2894.8321, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 35: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 13: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 15: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 16: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 40: [2460.5101999999997, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 21: [1667.3942, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 38: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 46: [2864.9471, 0.0]
[info | PowerModels]: removing 1 cost terms from generator 45: [1261.217, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 36: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 48: Float64[]
[info | PowerModels]: removing 3 cost terms from generator 18: Float64[]
[info | PowerModels]: removing 1 cost terms from generator 30: [2575.8442, 0.0]
[info | PowerModels]: removing 3 cost terms from generator 3: Float64[]
[info | PowerModels]: updated generator 32 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 29 cost function with order 2 to a function of order 3: [0.0, 3266.8781000000004, 0.0]
[info | PowerModels]: updated generator 1 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 54 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 2 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 41 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 51 cost function with order 2 to a function of order 3: [0.0, 3504.3401000000003, 0.0]
[info | PowerModels]: updated generator 53 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 27 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 42 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 33 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 28 cost function with order 2 to a function of order 3: [0.0, 3478.1778000000004, 0.0]
[info | PowerModels]: updated generator 50 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 52 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 26 cost function with order 2 to a function of order 3: [0.0, 1605.6042000000002, 0.0]
[info | PowerModels]: updated generator 10 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 24 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 25 cost function with order 2 to a function of order 3: [0.0, 2486.1868, 0.0]
[info | PowerModels]: updated generator 23 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 49 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 5 cost function with order 2 to a function of order 3: [0.0, 2498.342, 0.0]
[info | PowerModels]: updated generator 31 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 43 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 39 cost function with order 2 to a function of order 3: [0.0, 3407.2633000000005, 0.0]
[info | PowerModels]: updated generator 34 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 44 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 17 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 37 cost function with order 2 to a function of order 3: [0.0, 2460.0771999999997, 0.0]
[info | PowerModels]: updated generator 47 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 9 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 12 cost function with order 2 to a function of order 3: [0.0, 2222.098, 0.0]
[info | PowerModels]: updated generator 20 cost function with order 2 to a function of order 3: [0.0, 2420.2306, 0.0]
[info | PowerModels]: updated generator 6 cost function with order 2 to a function of order 3: [0.0, 12458.1564, 0.0]
[info | PowerModels]: updated generator 14 cost function with order 2 to a function of order 3: [0.0, 2599.3982, 0.0]
[info | PowerModels]: updated generator 7 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 8 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 19 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 4 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 22 cost function with order 2 to a function of order 3: [0.0, 2727.7342999999996, 0.0]
[info | PowerModels]: updated generator 11 cost function with order 2 to a function of order 3: [0.0, 2894.8321, 0.0]
[info | PowerModels]: updated generator 35 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 13 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 15 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 16 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 40 cost function with order 2 to a function of order 3: [0.0, 2460.5101999999997, 0.0]
[info | PowerModels]: updated generator 21 cost function with order 2 to a function of order 3: [0.0, 1667.3942, 0.0]
[info | PowerModels]: updated generator 38 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 46 cost function with order 2 to a function of order 3: [0.0, 2864.9471, 0.0]
[info | PowerModels]: updated generator 45 cost function with order 2 to a function of order 3: [0.0, 1261.217, 0.0]
[info | PowerModels]: updated generator 36 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 48 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 18 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
[info | PowerModels]: updated generator 30 cost function with order 2 to a function of order 3: [0.0, 2575.8442, 0.0]
[info | PowerModels]: updated generator 3 cost function with order 0 to a function of order 3: [0.0, 0.0, 0.0]
ERROR: BoundsError: attempt to access SymbolicUtils.BasicSymbolic{Array{Real, 3}} at index [56, 40, 41]
Stacktrace:
 [1] getindex(::SymbolicUtils.BasicSymbolic{Array{Real, 3}}, ::Int64, ::Int64, ::Int64)
   @ Symbolics ~/.julia/packages/Symbolics/k84uz/src/array-lib.jl:27
 [2] getindex(::Symbolics.Arr{Symbolics.Num, 3}, ::Int64, ::Int64, ::Int64)
   @ Symbolics ~/.julia/packages/Symbolics/k84uz/src/array-lib.jl:90
 [3] solve_opf(file_name::String)
   @ Main /private/tmp/mtk.jl:37
 [4] top-level scope
   @ REPL[9]:1
```

**Comment by Vaibhavdixit02** [comment]
>On the 118 bus case I get

Works now, the `p` and `q` variables were hardcoded to 1:6 that was the cause of the error.

>How does this do on say, the 118 case here: https://github.com/lanl-ansi/rosetta-opf/blob/main/data/pglib_opf_case118_ieee.m?

Has been running for past 10mins on my laptop and hasn't finished yet, will report the run time when it finishes

File: mtk.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    ModelingToolkit.@variables vm[keys(ref[:bus])] = ones(length(ref[:bus])) (bounds = [(ref[:bus][i]["vmin"], ref[:bus][i]["vmax"]) for i in keys(ref[:bus])])
+    vars = vcat(vars, [va[i] for i in keys(ref[:bus])], [vm[i] for i in keys(ref[:bus])])
+    ModelingToolkit.@variables pg[keys(ref[:gen])] (bounds = [(ref[:gen][i]["pmin"], ref[:gen][i]["pmax"]) for i in keys(ref[:gen])])
+    ModelingToolkit.@variables qg[keys(ref[:gen])] (bounds = [(ref[:gen][i]["qmin"], ref[:gen][i]["qmax"]) for i in keys(ref[:gen])])
+    vars = vcat(vars, [pg[i] for i in keys(ref[:gen])], [qg[i] for i in keys(ref[:gen])])
+    i_inds, j_inds, l_inds = maximum(first.(ref[:arcs])), maximum(getindex.(ref[:arcs], Ref(2))), maximum(last.(ref[:arcs]))
+    ModelingToolkit.@variables p[1:i_inds, 1:j_inds, 1:l_inds]
+    ModelingToolkit.@variables q[1:i_inds, 1:j_inds, 1:l_inds]
+
+    for (l, i, j) in ref[:arcs]
```
Comment (odow) [inline mtk.jl:37]: Ah. I guess the problem is that you're generating a dense matrix for `p` and `q`, but the `vars` is highly sparse. You might be better off with a dictionary mapping `(l, i, j)` to a variable?

**Comment by odow** [comment]
> Has been running for past 10mins on my laptop and hasn't finished yet

You might want to try a smaller case then. Here's a 14 bus case (and there are all the others in that repo):

https://github.com/power-grid-lib/pglib-opf/blob/master/pglib_opf_case14_ieee.m

A a frame of reference, we're aiming to solve the 118 case in less than one second:

image

File: mtk.jl (hunk 1 above)
Comment (Vaibhavdixit02) [reply mtk.jl:37]: Ah okay, I was thinking about this. I'll try this

File: mtk.jl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    for (i, bus) in ref[:ref_buses]
+        push!(cons, va[i] ~ 0)
+    end
+
+    for (i, bus) in ref[:bus]
+        bus_loads = [ref[:load][l] for l in ref[:bus_loads][i]]
+        bus_shunts = [ref[:shunt][s] for s in ref[:bus_shunts][i]]
+        push!(cons,
+            sum(p[a...] for a in ref[:bus_arcs][i]) ~
+                (isempty(ref[:bus_gens][i]) ? 0.0 : sum(pg[g] for g in ref[:bus_gens][i])) -
```
Comment (odow) [inline mtk.jl:54]:
```suggestion
                sum(pg[g] for g in ref[:bus_gens][i]; init = 0.0) -
```
Perhaps? I didn't try locally

File: mtk.jl [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                (isempty(bus_loads) ? 0.0 : sum(load["pd"] for load in bus_loads))
+            # sum(shunt["gs"] for shunt in bus_shunts)*vm[i]^2
```
Comment (odow) [inline mtk.jl:56]: I guess this is still a TODO

**Comment by odow** [comment]
@Vaibhavdixit02 perhaps we can find time at JuliaCon to discuss/finish this.

@ccoffrin I'll leave this one open for now.

**Comment by Vaibhavdixit02** [comment]
@odow we should do that!

**Comment by ccoffrin** [comment]
Sounds good to me.  Knowing almost nothing about the SciML package ecosystem, I am curious to know more about what is expected to be the difference between an implementation in Optimization.jl and ModelingToolkit.jl

**Comment by Vaibhavdixit02** [comment]
This should be ready now. Further updates if any will be at the packages and the script will not have to change.

I am curious which time is the one that's reported in the tables on discourse? In this implementation I expect huge (really huge) compile times because of the symbolic derivative creations.

I plan to create another PR to update the Optimization.jl implementation using SparseDiffTools (due to nested finite difference it gives some inaccurate hessians) and one of the reverse mode AD packages - I realised I haven't added constraints support for either zygote or reversediff and enzyme doesn't yet work on the script

**Comment by ccoffrin** [comment]
I did some testing and there seems to be an issue with correctness of the implementation.  This is what I found in the two datasets that are provided with the repo,
- `pglib_opf_case5_pjm.m`  final objective should be 17551.9 but is 17929.2
- `pglib_opf_case118_ieee.m` took too long to run for me to test locally

```
  [b6b21f68] Ipopt v1.4.1
  [961ee093] ModelingToolkit v8.64.0
  [7f7a1694] Optimization v3.15.2
  [fd9f6733] OptimizationMOI v0.1.15
  [c36e90e8] PowerModels v0.19.9
```

From a quick skim of the model, the only thing that stood out to be as obviously inconsistent was this,
```
push!(cons, abs(va_fr - va_to) ≲ branch["angmax"])
```
Should be something that captures two constraints like so,
```
 branch["angmin"] <= va_fr - va_to <= branch["angmax"]
```
No that it may be the case that ("angmin") is not the same as (-"angmax").

Regarding your other question, the time reported in the table on discourse is the "total time" value. This seems to be the most appropriate across modeling tools because some tools do some pre-compilation during model building while others wait until the final "solve" call to do similar work. "total time" ensures all aspects are captured. In preparing the data for the table I do run one small case before collecting timing information to remove Julia's JIT related overheads.

File: mtk.jl [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        push!(cons, p_to^2 + q_to^2 ≲ branch["rate_a"]^2)
+    end
+
+
+    model_build_time = time() - time_model_start
+
+
+    time_solve_start = time()
+    optsys = ModelingToolkit.OptimizationSystem(loss, vars, [], constraints=cons, name=:rosetta)
+    optsys = ModelingToolkit.structural_simplify(optsys)
```
Comment (Vaibhavdixit02) [inline mtk.jl:138]:
```suggestion
```

**Comment by Vaibhavdixit02** [comment]
I thought of updating this but might be cleaner to just copy the mtk one from scimlbenchmarks here, with disabling `structural_simplify` though

**Comment by Vaibhavdixit02** [comment]
Sure. 

MTK is a general purpose CAS and provides a symbolic interface for optimization problems as well. The way it connects with Optimization.jl is by relying on the same interface to solvers as the functional interface provided by Optimization.jl. You could think of it as a setup similar to JuMP and MOI, but here both being more higher level right now (since we interface with a lot of the the solvers only through MOI as well). MTK generates its own symbolic derivatives and uses those oracles while Optimization.jl provides the ability to choose all the available AD choices (that it supports currently), as well as symbolic derivatives by using the same system as is used when you create the model with MTK but through the functional interface for the user ("just write julia functions")  via the `AutoModelingToolkit` AD choice which is what we used here in the earliest implementations.

I hope I didn't confuse you more! 😅

**Comment by ccoffrin** [comment]
Thanks that does help. Based on your description here it does seem that a ModelingToolkit variant could be a nice addition to the modeling layer examples in the root directory, unless you strongly disagree.

From a quick glance at SciMLBenchmarks I did not see the MTK implementation, can you point me to it more specifically?

**Pull request closed without merging**